Reuse HttpContext object per HTTP/1 connection and HTTP/2 stream

davidfowl opened the pull request:
- Today in Kestrel, we reuse the IFeatureCollection per Http1Connection and per Http2Stream. This PR aims to take advantage of that same technique and affinitize the HttpContext and friends so that they are only allocated once per connection.

cc @benaadams 

PS: I looked at doing this in the default HttpContextFactory impl with a feature based approach but there's no place to store the HttpContext that isn't per request in other servers.

I'll try to get some perf results for allocations and throughput as well.

Here's a profile before all of the optimizations we've done (including a bunch from @benaadams):
- Logging is off in this benchmark so there's are no AsyncLocal or ExecutionContext allocations
- I turned off tiered compilation to reduce the noise (some JIT optimizations are turned off in tier0)

![image](https://user-images.githubusercontent.com/95136/50731876-c483a480-1123-11e9-8e63-8704915e8e89.png)

I'm still waiting on https://github.com/dotnet/coreclr/issues/21559 to be fixed to get an equivalent profile. Here's what I expect allocation wise:
- We'll get rid of HttpContext, HttpRequest and HttpResponse allocations as part of this PR
- The QueueUserWorkItem callbacks should be completely gone by now (latest coreclr (https://github.com/dotnet/coreclr/commit/e7ead79fedc52e17f2cf9befd5c0f5091d70f909#diff-b9b22fdb097c803c0ca6fa45655b24f7) plus the IThreadPoolWorkItem impl on IOQueue https://github.com/aspnet/AspNetCore/commit/0d9de4991eaae9401d64cd2eafa42f6aebde33ae)
- The `byte[]` allocations were cut in 1/2 because of this https://github.com/aspnet/AspNetCore/commit/cb1917aa5990a701b92fc44f28f76b0718c76197 and will be fully eliminated once we do the pipelines transition (we'll be using Kestrel buffers instead of allocating temporary ones).
- The HttpMethod boxing is fixed by this https://github.com/aspnet/AspNetCore/commit/8be8e0008dec5af5b73888b48d65d5bfc19cdf62
- This will also work for HTTP/2 once we implement the pooling as described here https://github.com/aspnet/AspNetCore/issues/6192

That leaves us with the following allocations:
- RequestServicesFeature (no plan yet)
- FormFeature (https://github.com/aspnet/HttpAbstractions/pull/999/files)
- String (header allocations) (potentially make lazy)

Here's what I got when looking at allocations with dotTrace's timeline view (thanks @pakrym)

## Before

![image](https://user-images.githubusercontent.com/95136/50747826-e3993980-11ea-11e9-86f4-b362a094db1f.png)

## After

![image](https://user-images.githubusercontent.com/95136/50747499-0d516100-11e9-11e9-94c4-2f436b15c142.png)

We removed 35MB from the small object heap with this change.

### Strings

As for the strings It's all headers and path:

![image](https://user-images.githubusercontent.com/95136/50747564-59040a80-11e9-11e9-8d8c-38f0664e562d.png)

davidfowl committed:
Reuse HttpContext object per connection.
- Today in Kestrel, we reuse the IFeatureCollection per connection and per Http2Stream. This PR aims to take advantage of that same technique and affinitize the HttpContext and friends so that they are only allocated per connection.
- ReusableHttpContext and friends mimic the functionality of DefaultHttpContext but is sealed and has no overridable methods.

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/KestrelHttpContextFactory.cs:
```diff
@@ -0,0 +1,55 @@
+
+        public override void Dispose(HttpContext httpContext)
+        {
+            if (httpContext is DefaultHttpContext defaultHttpContext)
```
This isn't strictly necessary and in fact will lead to null refs today. We should make Uninitiaize throw an better exception when GetFeature is called on the FeatureCollection (but it's currently being nulled out).

davidfowl commented on src/Http/Http/src/DefaultHttpContext.cs:
```diff
@@ -51,13 +51,92 @@ public DefaultHttpContext(IFeatureCollection features)
-            _features = default(FeatureReferences<FeatureInterfaces>);
+            _features = default;
+
+            if (GetType() != typeof(DefaultHttpContext))
```
I have to see how it affects the performance also.

halter73 replied:
One easy way to improve the perf is to move this condition to the DefaultHttpContext constructor to be cached into an _isDerived field. Additionally, you could add an internal constructor that's only called by HttpContextFactory that always sets _isDerived to false. That way, in the default case, there would be no calls to GetType().

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -535,14 +540,24 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
 
-                var httpContext = application.CreateContext(this);
+                // Initialize the HttpContext before we call into the IHttpApplication
+                if (_httpContext is null)
```
If this is a step too far I can revert this commit and go back to this one https://github.com/aspnet/AspNetCore/pull/6424/commits/8526622fefd26f95528ecf9127ce6a9b7c7f1315. The difference is that it would lazily initialize the http context in the rare case where the IHttpContextFactory is overridden by the user. It means this field would always be null instead of being assigned and those extra calls would be avoided.

sebastienros commented:
Before: 1,927,129
After: 1,847,978

4% regression

```
dotnet run -- --server http://asp-perf-win:5001 --client http://asp-perf-load:5002 `
-j ../Benchmarks/benchmarks.plaintext.json -n Plaintext `
--build-id 67321 --include Microsoft.AspNetCore.Http.* --include Microsoft.AspNetCore.Server.Kestrel.* `
--self-contained -i 5 -x 1
```

halter73 commented on src/Http/Http/src/Internal/ReusableHttpContext.cs:
```diff
@@ -0,0 +1,195 @@
+        }
+
+
+
```
Nit: newlines

halter73 commented on src/Http/Http/src/Internal/ReusableHttpContext.cs:
```diff
@@ -0,0 +1,195 @@
+        /// This is obsolete and will be removed in a future version. 
+        /// The recommended alternative is to use Microsoft.AspNetCore.Authentication.AuthenticationHttpContextExtensions.
+        /// See https://go.microsoft.com/fwlink/?linkid=845470.
+        /// </summary>
```
Will anyone see doc comments on this type?

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -275,6 +277,8 @@ public CancellationToken RequestAborted
 
         protected HttpResponseHeaders HttpResponseHeaders { get; } = new HttpResponseHeaders();
 
+        internal HttpContext HttpContext => _httpContext;
```
Why not use an auto-property?

halter73 commented on src/Http/Http/src/HttpContextFactory.cs:
```diff
@@ -47,6 +47,11 @@ public HttpContext Create(IFeatureCollection featureCollection)
 
+        protected virtual HttpContext CreateHttpContext(IFeatureCollection featureCollection)
+        {
+            return new DefaultHttpContext(featureCollection);
```
Might as well use the ReusableHttpContext here too.

davidfowl commented:
```
RequestsPerSecond:           1,920,916
```

Tratcher commented on src/Servers/Kestrel/Kestrel/src/WebHostBuilderKestrelExtensions.cs:
```diff
@@ -32,6 +33,7 @@ public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder)
                 services.TryAddSingleton<ITransportFactory, SocketTransportFactory>();
 
                 services.AddTransient<IConfigureOptions<KestrelServerOptions>, KestrelServerOptionsSetup>();
+                services.AddSingleton<IHttpContextFactory, KestrelHttpContextFactory>();
```
So what happens when IIS in-proc implements the same feature?

davidfowl replied:
What are you asking? When happens if somebody replaces this feature? Kestrel will work but will be un-optimized.

davidfowl replied:
I can test the performance of the interface approach and make changes to the default factory. That would be a bit more cooperative.

Tratcher reviewed (commented):
Not a fan of the blanket copy paste, it creates a long term maintenance problem. You can't derive from DefaultHttpContext and then call base, right?

davidfowl commented:
> Not a fan of the blanket copy paste, it creates a long term maintenance problem. You can't derive from DefaultHttpContext and then call base, right?

That defeats the purpose. This change effectively makes DefaultHttpContext legacy without breaking existing consumers of it.

davidfowl committed:
Introduce IHttpContextContainer
- Allows servers to cache the HttpContext and friends across requests.
- Remove KestrelHttpContextFactory

davidfowl commented:
@Tratcher 🆙 📅

Tratcher commented on src/Servers/Kestrel/Kestrel/src/WebHostBuilderKestrelExtensions.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Http;
```
Revert

davidfowl committed:
More cleanup

davidfowl commented on src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs:
```diff
@@ -275,6 +277,23 @@ public CancellationToken RequestAborted
+        {
+            get
+            {
+                if (_httpContext is null)
```
Hmm this change might be a bit dangerous if this is accessed during the request, it'll re-initialize?

halter73 replied:
Do you mean if it's accessed outside of the request flow?

davidfowl replied:
Inside. If you get the HttpContext and cast the FeatureCollection property to IHttpContextContainer then it’ll call Initialize again

benaadams replied:
Create it upfront in the variable
```
private ReusableHttpContext _httpContext = new ReusableHttpContext(this);
```
Add an `IsInitialized` flag in `ReusableHttpContext` and use that in the property?

halter73 replied:
I'm not sure I understand what you're getting at. You mean something like the following?

```C#
app.Run(async context =>
{
    var context2 = ((IHttpContextContainer)context.Features).HttpContext;
    await context.Response.WriteAsync(object.ReferenceEquals(context, context2).ToString());
});
```

In that case the response would be "True". I'm not sure why you'd think this wouldn't be the case.

I'm also not sure why we're concerned about what happens when the IFeatureCollection to IHttpContextContainer. I don't know why we'd have any burden whatsoever to support that.

davidfowl replied:
It's super edge case. I'm going to merge this anyways, I just wanted to bring it up.
